std/range/primitives.d

```diff
@@ -123,16 +123,28 @@ r.popFront();     // can invoke popFront()
 auto h = r.front; // can get the front of the range of non-void type
 ----
 
-The semantics of an input range (not checkable during compilation) are
-assumed to be the following ($(D r) is an object of type $(D R)):
-
-$(UL $(LI $(D r.empty) returns $(D false) iff there is more data
-available in the range.)  $(LI $(D r.front) returns the current
-element in the range. It may return by value or by reference. Calling
-$(D r.front) is allowed only if calling $(D r.empty) has, or would
-have, returned $(D false).) $(LI $(D r.popFront) advances to the next
-element in the range. Calling $(D r.popFront) is allowed only if
-calling $(D r.empty) has, or would have, returned $(D false).))
+The following are rules of input ranges are assumed to hold true in all
+Phobos code. These rules are not checkable at compile-time, so not conforming
+to these rules when writing ranges or range based code will result in
+undefined behavior.
+
+$(UL
+    $(LI `r.empty` returns `false` iff there is more data
+    available in the range.)
+    $(LI `r.empty` called multiple times, without calling
+    `r.popFront`, or otherwise mutating the range object,
+    yields the same result for every call.)
```

Comment:
I don't like the term "calling", because it implies that `empty` must be a function. That's not necessarily the case, since it can be a member variable of the range that gets updated by `popFront`. Maybe "evaluated" is a better word?

Also, multiple evaluations of `empty` without calling `popFront` should not cause the value of `front` to change. Or is this already covered by the following rules?